File: src/norswap/autumn/parsers/Lookahead.java
package norswap.autumn.parsers;

import norswap.autumn.Parse;
import norswap.autumn.Parser;
import norswap.autumn.ParserVisitor;
import java.util.Collections;

/**
 * Succeeds if its child succeed, but does not advance the input position (all other side effects
 * of the child are retained).
 *
 * <p>Build with {@link norswap.autumn.Grammar.rule#ahead()}
 */
public final class Lookahead extends Parser
{
    // ---------------------------------------------------------------------------------------------

    public final Parser child;

    // ---------------------------------------------------------------------------------------------

    public Lookahead (Parser child)
    {
        this.child = child;
    }

    // ---------------------------------------------------------------------------------------------

    @Override public boolean doparse (Parse parse)
    {
        int pos0 = parse.pos;
        if (child.parse(parse)) {
            parse.pos = pos0;
            return true;
        }
        // parse.pos has already been reset
        return false;
    }

    // ---------------------------------------------------------------------------------------------

    @Override public void accept (ParserVisitor visitor) {
        visitor.visit(this);
    }

    // ---------------------------------------------------------------------------------------------

    @Override public Iterable<Parser> children() {
        return Collections.singletonList(child);
    }

    // ---------------------------------------------------------------------------------------------

    @Override public String toStringFull() {
        return "lookahead(" + child + ")";
    }

    // ---------------------------------------------------------------------------------------------
}

File: src/norswap/autumn/parsers/Lookahead.kt
package norswap.autumn.parsers
import norswap.autumn.Grammar
import norswap.autumn.Parser
import norswap.autumn.BadMatch

// -------------------------------------------------------------------------------------------------

/**
 * Succeeds if [p] succeeds, but does not advance the input position (all other side effects of
 * [p] are retained).
 */
inline fun Grammar.ahead (crossinline p: Parser): Boolean
{
    val pos0 = pos
    val result = ignore_errors_if_successful(p)
    pos = pos0
    return result
}

// -------------------------------------------------------------------------------------------------

/**
 * Succeeds if [p] succeeds, but does produce any side effect (does not even change the input
 * position).
 */
inline fun Grammar.ahead_pure (crossinline p: Parser): Boolean
{
    val pos0 = pos
    val ptr0 = log.size
    val result = ignore_errors_if_successful(p)
    undo(pos0, ptr0)
    return result
}

// -------------------------------------------------------------------------------------------------

/**
 * Succeeds only if [p] fails.
 */
inline fun Grammar.not (crossinline p: Parser): Boolean
{
    val pos0 = pos
    val ptr0 = log.size
    val result = ignore_errors(p)
    if (result) {
        undo(pos0, ptr0)
        fail(pos0, BadMatch)
    }
    return !result
}

// -------------------------------------------------------------------------------------------------
